go/arrow/avro/reader_types.go

```diff
@@ -0,0 +1,892 @@
+}
+
+func (f *fieldPos) children() []*fieldPos { return f.childrens }
+
+func (f *fieldPos) metadata() arrow.Metadata { return f.metadatas }
+
+func (f *fieldPos) newChild(childName string, childBuilder array.Builder, meta arrow.Metadata) *fieldPos {
+	var child fieldPos = fieldPos{
+		parent:    f,
+		fieldName: childName,
+		builder:   childBuilder,
+		metadatas: meta,
+		index:     int32(len(f.childrens)),
+		depth:     f.depth + 1,
+	}
+	if f.isList {
+		child.isItem = true
+	}
+	child.path = child.namePath()
+	f.childrens = append(f.childrens, &child)
+	return &child
+}
+
+// NamePath returns a slice of keys making up the path to the field
+func (f *fieldPos) namePath() []string {
+	if len(f.path) == 0 {
+		var path []string
+		var listPath []string
+		cur := f
+		for i := f.depth - 1; i >= 0; i-- {
+			if cur.typeName == "" {
+				path = append([]string{cur.fieldName}, path...)
+			} else {
+				path = append([]string{cur.fieldName, cur.typeName}, path...)
+			}
+			if !cur.parent.isMap {
+				cur = cur.parent
+			}
+		}
+		if f.parent.parent != nil && f.parent.parent.isList {
+			for i := len(path) - 1; i >= 0; i-- {
+				if path[i] != "item" {
+					listPath = append([]string{path[i]}, listPath...)
+				} else {
+					return listPath
+				}
+			}
+		}
+		if f.parent != nil && f.parent.fieldName == "value" {
+			for i := len(path) - 1; i >= 0; i-- {
+				if path[i] != "value" {
+					listPath = append([]string{path[i]}, listPath...)
+				} else {
+					return listPath
+				}
+			}
+		}
+		return path
+	}
+	return f.path
+}
+
+// GetValue retrieves the value from the map[string]any
+// by following the field's key path
+func (f *fieldPos) getValue(m any) any {
+	var value any = m
+	// The value of the type assertion is not assigned to a variable as `value` is reassigned
+	// and its type reasserted as the data is traversed
+	switch value.(type) {
+	case map[string]any:
+		for _, key := range f.namePath() {
+			valueMap, ok := value.(map[string]any)
+			if !ok {
+				if key == "item" {
+					return value
+				}
+				return nil
+			}
+			value, ok = valueMap[key]
+			if !ok {
+				return nil
+			}
+		}
+	default:
+		return value
+	}
+	return value
+}
+
+// Avro data is loaded to Arrow arrays using the following type mapping:
+//
+//	Avro					Go    			Arrow
+//	null					nil				Null
+//	boolean					bool			Boolean
+//	bytes					[]byte			Binary
+//	float					float32			Float32
+//	double					float64			Float64
+//	long					int64			Int64
+//	int						int32  			Int32
+//	string					string			String
+//	array					[]interface{}	List
+//	enum					string			Dictionary
+//	fixed					[]byte			FixedSizeBinary
+//	map and record	map[string]any	Struct
+//
+// mapFieldBuilders builds a tree of field builders matching the Arrow schema
+func mapFieldBuilders(b array.Builder, field arrow.Field, parent *fieldPos) {
+	switch bt := b.(type) {
+	case *array.BinaryBuilder:
+		f := parent.newChild(field.Name, bt, field.Metadata)
```

Comment:
it looks like this line is in every case, right? can we hoist it out above the switch instead?

Reply:
yup, done